libcontainer/state_linux.go

```diff
@@ -88,6 +88,15 @@ func (b *stoppedState) transition(s containerState) error {
 }
 
 func (b *stoppedState) destroy() error {
+	// Normally, when a container init is gone, all other processes in its
+	// cgroup are killed by the kernel. This is not the case for a shared
+	// PID namespace container, which may have some leftover processes.
+	if !b.c.config.Namespaces.IsPrivate(configs.NEWPID) {
+		// Get container pids on a best-effort basis.
+		if pids, _ := b.c.cgroupManager.GetAllPids(); len(pids) > 0 {
+			return fmt.Errorf("container is in stopped state (init is gone), but its cgroup still has %d processes; use runc delete -f or runc kill", len(pids))
+		}
+	}
```

Comment:
What you mean is that we should drop this part, change to kill the left processes in the stopped container when using `runc delete` without `-f`?

Reply:
Yes, I think `runc delete` on a container with a dead init process should work without `-f` because the container is considered "stopped".

Reply:
I can accept to change back the same as before.

But there is another issue discussed in https://github.com/opencontainers/runc/issues/4040#issuecomment-1752341095 (I can accept fix it in another PR, this PR is only to fix the regression introduced by https://github.com/opencontainers/runc/pull/3825)
Maybe we have to retry and still have to check whether there are still some processes left in the container or not, even we had try our best to kill them. At this time, we should error out and don't delete anything about this container.
Please see: https://github.com/opencontainers/runc/pull/4048/commits/120dad41376af412c6f6c6dace0fbd7f2b39b42e

Reply:
Ah, I see. I wonder if @kolyshkin's suggestion in https://github.com/opencontainers/runc/issues/4040#issuecomment-1743983574 is workable. @kolyshkin is there a reason you didn't go the route of making `!c.hasInit()` into `!c.hasProcesses()`, so that the container is still in the `running` state despite not having a pid1?

My issue with returning an error here is that it makes `stopped` containers act strangely in this one scenario. If we treated containers that contain processes _at all_ as `running` then we can require people to use `runc kill`, which would solve the consistency issue.

Reply:
I will change the last commit to kill the remaining processes rather than error out.

Reply:
done